[SYCL][ESIMD] Make logs architecture more flexible

yuriykoch opened the pull request:
Removes tight coupling of the test case description and error message, so that
we could to use any error messages with no need of manual message construction.
Any log message specific details are moved to the upper level, to the message
construction itself, because log messages are entry points for test failure
analysys.
Also a syntax sugar for trace and debug messages added.

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Make logs architecture more flexible

Removes tight coupling of the test case description and error message, so that
we could to use any error messages with no need of manual message construction.
Any log message specific details are moved to the upper level, to the message
construction itself, because log messages are entry points for test failure
analysys.
Also a syntax sugar for trace and debug messages added.

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
Hotfix types for operator_assignment_move_and_copy_core

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Minimize TestCaseDescription template parameters

To boost up compilation performance a bit by avoiding unneccessary
instantiations

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Add floating point hex values to logs

Hex representation:
- states differences in 1 ULP explicitly
- makes it easier to reproduce any precision-related issue
- provides additional information regarding NaN values

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
Merge branch 'intel' into feature/logs

vasilytric commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -14,36 +14,26 @@
-  TestDescription(size_t index, DataT retrieved_val, DataT expected_val,
-                  const std::string &data_type)
-      : m_data_type(data_type), m_retrieved_val(retrieved_val),
-        m_expected_val(expected_val), m_index(index) {}
+  TestDescription(const std::string &data_type) : m_data_type(data_type) {}
```
Why you do not provide a backward compatibility? Without the backward compatibility new tests will be failed on CI when this PR will be merged

please leave the previous version of the `TestDescription` constructor without any changed in `to_string` method

yuriykoch replied:
Would you mind to take a look on a backward compatibility suggestion? This PR removes unused `DataT` template parameter, so we have a difference in template parameters also.

yuriykoch committed:
[SYCL][ESIMD] Add backward compatibility for logs

The feature flag looks like the best option to avoid ODR violation for
different template parameters in this case:
- both default value and parameter pack are not applicable
- using different specializations of primary template:

    template <typename ...> struct TestDescription;
    template <typename T, int N>
    struct [[deprecated]] TestDescription<T, std::integral_constant<int,N>>{
      // deprecated implementation
    };
    template <int N>
    struct TestDescription<std::integral_constant<int,N>>{
      // new implementation
    };

  wouldn't work without switching to integral_constant, which seems like
  not the best option in this case

yuriykoch committed:
Merge branch 'intel' into feature/logs

yuriykoch committed:
Merge branch 'intel' into feature/logs

With updates moved from ctor_load_core.cpp to ctor_load.hpp

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -14,20 +14,44 @@
+//
+// TODO: Remove deprecated TestDescription from all tests
+template <typename DataT, int NumElems, typename ContextT>
+class [[deprecated]] TestDescription : public ITestDescription {
```
@vasilytric This attribute would trigger a compiler warning for every template instantiation by-design. Let me know if that's not needed from your point of view.

vasilytric replied:
I think that this attribute will be useful

yuriykoch commented:
@sndmitriev Would you mind to take a look? It's a generic code, so I'm going to support this branch in case any other test is merged - still there is no any dependency in merge order with other PRs.

yuriykoch committed:
Merge branch 'intel' into feature/logs

with log message for ctor_default tests updated as a part
of merge conflict resolution

yuriykoch committed:
[SYCL][ESIMD] Fix backward compatibility for `log::note`

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch commented:
Left the `operator_logical_not` test untouched, with deprecated logs enabled. Looks like the log messages update there could trigger code refactoring out of scope of this PR due to equal error messages for different logic.
Tagging @vasilytric for awareness

yuriykoch committed:
Merge branch 'intel' into feature/logs

yuriykoch committed:
[SYCL][ESIMD] Update `should_skip_test_with<T>` logs

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYC}[ESIMD] Update `operator_decrement_and_increment` tests

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch commented:
@sndmitriev Friendly ping

yuriykoch commented:
@kbobrovs Would you mind to merge this PR?
